assets/javascripts/discourse/components/query-row-content.js

```diff
@@ -9,62 +9,18 @@ import { get } from "@ember/object";
 import { isEmpty } from "@ember/utils";
 import { escapeExpression } from "discourse/lib/utilities";
 
-function icon_or_image_replacement(str, ctx) {
-  str = get(ctx.contexts[0], str);
-  if (isEmpty(str)) {
-    return "";
-  }
-
-  if (str.indexOf("fa-") > -1) {
-    const icon = iconHTML(convertIconClass(str));
-    return htmlSafe(icon);
-  } else {
-    return htmlSafe("<img src='" + str + "'>");
-  }
-}
-
-function category_badge_replacement(str, ctx) {
-  const category = get(ctx.contexts[0], str);
-  return categoryLinkHTML(category, {
-    allowUncategorized: true,
-  });
-}
+export default class QueryRowContent extends Component {
+  constructor() {
+    super(...arguments);
 
-function bound_date_replacement(str, ctx) {
-  const value = get(ctx.contexts[0], str);
-  return htmlSafe(autoUpdatingRelativeAge(new Date(value), { title: true }));
-}
-
-// consider moving this elsewhere
-function guessUrl(t) {
-  let [dest, name] = [t, t];
-
-  const split = t.split(/,(.+)/);
-
-  if (split.length > 1) {
-    name = split[0];
-    dest = split[1];
-  }
-
-  return [dest, name];
-}
-
-const QueryRowContentComponent = Component.extend({
-  tagName: "tr",
-  rowContents: null,
-
-  didReceiveAttrs() {
-    const row = this.row;
-    const parentView = this.parentView;
-    const fallback = this.fallbackTemplate;
     const helpers = {
       "icon-or-image": icon_or_image_replacement,
       "category-link": category_badge_replacement,
       reltime: bound_date_replacement,
     };
 
-    const parts = this.columnTemplates.map((t, idx) => {
-      const value = row[idx],
+    this.results = this.args.columnTemplates.map((t, idx) => {
```

Comment:
If any arguments change after the initial render, `results` is not going to be recomputed. I guess a general rule-of-thumb to avoid this is "don't use `this.args` inside the constructor".

Previously, `didReceiveAttrs` was re-run on every argument change, so this *might* be a regression (although it obviously depends on precisely how the component is used)

Could we make this a cached getter instead?

```javascript
@cached
get results(){
  return this.args.columnTemplates.map(...
```

Reply:
Great tip, thank you. Change added.